### PR TITLE
Add scipy and numdifftools Hessian approximations

### LINKED ISSUE #942: Add finite difference Hessians
We should add scipy and numdifftools finite difference Hessian approximations.

### PR DESCRIPTION
**Pull request by jfowkes**
#### Description of Work

Resolve #942

#### Testing Instructions

1.
2.
3.

Function: Does the change do what it's supposed to?

Tests: Does it pass? Is there adequate coverage for new code?

Style: Is the coding style consistent? Is anything overly confusing?

Documentation: Is there a suitable change to documentation for this change?

### CODE REVIEW

**Comment by jfowkes** [comment]
Failing due to #944

**Comment by jfowkes** [comment]
This is now ready to be reviewed. Note that I have fixed the docs to say that only the `cutest` parser supports analytic Jacobians and Hessians as is the case in the code (or at least in `options.py`). Linting tests are only failing due to the `test_base_table.py` errors (nothing to do with this PR).

File: fitbenchmarking/utils/options.py [hunk 1]
```diff
@@ -340,9 +360,12 @@ def list_to_string(mylist):
                                  self.algorithm_type),
                              'software': list_to_string(self.software),
                              'jac_method': list_to_string(self.jac_method),
+                             'hes_method': list_to_string(self.hes_method),
                              'max_runtime': self.max_runtime}
         config['JACOBIAN'] = {k: list_to_string(m)
                               for k, m in self.num_method.items()}
```
Comment (AndrewLister-STFC) [inline fitbenchmarking/utils/options.py:366]: I'm wondering if this should be renamed to `jac_num_method`?

**Review by AndrewLister-STFC — CHANGES_REQUESTED** [review]
I think there's been some confusion here:

- NIST problems do have analytic jacobians
- The options are wierd and confusing

As far as I can tell the 'cutest' option is always passed to the analytic jacobian and never used, so I think there are 2 approaches here. 1) Fix it to avoid confusion now, or 2) mimic the jacobian implementation and create an issue to avoid scope creep.

File: fitbenchmarking/hessian/scipy_hessian.py [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+    Implements SciPy finite difference approximations to the derivative
+    """
+
+    # Problem formats that are incompatible with certain Scipy Hessians
+    INCOMPATIBLE_PROBLEMS = {"cs": ["mantid"]}
+
+    def eval(self, params, **kwargs):
+        """
+        Evaluates Hessian of problem.eval_model, returning the value
+        sum_{i=1}^m (r)_i \nabla^2r_i(x)
```
[the remark this answers is not in the document]
Comment (jfowkes) [reply fitbenchmarking/hessian/scipy_hessian.py:22]: Yes I think this is a stale comment from the previous version, note that this stale comment is also present in `analytic_jacobian.py`. Will fix this.

File: fitbenchmarking/hessian/scipy_hessian.py [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+        :param params: The parameter values to find the Hessian at
+        :type params: list
+
+        :return: Approximation of the Hessian
+        :rtype: numpy array
+        """
+        x = kwargs.get("x", self.problem.data_x)
+        hes = np.zeros((len(params), len(params), len(x)))
+        for i in range(len(x)):  # NOTE: subsetting Jacobian is inefficient
+            def grad_i(params): return self.problem.jacobian(x, params)[i, :]
```
[the remark this answers is not in the document]
Comment (jfowkes) [reply fitbenchmarking/hessian/scipy_hessian.py:33]: Unfortunately not as SciPy's `approx_derivative` needs to evaluate the jacobian at several points which vary according to the finite difference implementation used (p+h,p-h,etc) and `approx_derivative` doesn't support 3D arrays (unlike numdifftools).

File: fitbenchmarking/hessian/base_hessian.py [hunk 4]
```diff
@@ -32,3 +37,23 @@ def eval(self, params, **kwargs):
         :rtype: numpy array
         """
         raise NotImplementedError
+
+    @property
+    def method(self):
```
[the remark this answers is not in the document]
Comment (jfowkes) [reply fitbenchmarking/hessian/base_hessian.py:42]: I just copied what was in `base_jacobian.py`

File: fitbenchmarking/hessian/tests/test_hessian.py [hunk 5]
```diff
@@ -187,6 +191,47 @@ def test_poisson(self):
...
+        """
+        for method in ['central',
+                       'forward',
+                       'backward',
+                       'complex',
+                       'multicomplex']:
+            hes = Numdifftools(self.cost_func.problem)
+            hes.method = method
+            eval_result = hes.eval(params=self.params)
+            self.assertTrue(np.isclose(self.actual_hessian, eval_result).all())
```
[the remark this answers is not in the document]
Comment (jfowkes) [reply fitbenchmarking/hessian/tests/test_hessian.py:233]: Yes I just copied what was in the Jacobian tests, happy to change but we'd have to change both Hessian and Jacobian tests.

File: fitbenchmarking/utils/options.py (hunk 1 above)
Comment (jfowkes) [reply fitbenchmarking/utils/options.py:366]: That would be a better name.

**Comment by jfowkes** [comment]
> I think there's been some confusion here:
> 
> * NIST problems do have analytic jacobians
> * The options are wierd and confusing
> 
> As far as I can tell the 'cutest' option is always passed to the analytic jacobian and never used, so I think there are 2 approaches here. 1) Fix it to avoid confusion now, or 2) mimic the jacobian implementation and create an issue to avoid scope creep.

Up to you really, there are a number of other inconsistencies in the Jacobian numerical method implementation and tests (and also in the Hessian as it's copied from the Jacobian) that you have raised above. Happy to deal with these here or in another issue.

File: fitbenchmarking/hessian/scipy_hessian.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (jfowkes) [reply fitbenchmarking/hessian/scipy_hessian.py]: Fixed in https://github.com/fitbenchmarking/fitbenchmarking/pull/943/commits/5ccc3e3b6cd1f819409ce8a7aee9a3b622d338c5

File: fitbenchmarking/hessian/scipy_hessian.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (AndrewLister-STFC) [reply fitbenchmarking/hessian/scipy_hessian.py:33]: Ahh yes I see. In that case, can we do something like:
```
for x_i in x:
    def grad_i(params): return self.problem.jacobian(x_i, params)
```

[the remark this answers is not in the document]
Comment (jfowkes) [reply fitbenchmarking/hessian/scipy_hessian.py:33]: Yes of course, why didn't I think of that?

**Comment by jfowkes** [comment]
@AndrewLister-STFC ready for another review, please take your time with this as much has changed.

**Review by AndrewLister-STFC — COMMENTED** [review]
Code looks good and the tests all pass 👍 

I'll update the branch, run it locally and then approve/merge

**Comment by AndrewLister-STFC** [comment]
Running this with RALFit seems to fail for all hessians (other than the default of None)
image
```
Error code from user-supplied subroutine eval_HF passed error = 2513
```

**Comment by AndrewLister-STFC** [comment]
Discussed with @jfowkes, 1 outstanding issue which is that hessians will always favour the analytic jacobian whereas they should use the numeric one if chosen

**Comment by jfowkes** [comment]
@AndrewLister-STFC I've spoken to Tyrone and RALFit requires `sum_{i=1}^m (r)_i \nabla^2r_i(x)` passed in as the Hessian. This used to be the case but it got forgotten about when we refactored. Now we could follow the cost functions on this and pass in:
```
r = self.cost_func.eval_r(params, **kwargs)
H, J = self.cost_func.hes_res(params, **kwargs)
return np.matmul(H, r))
```
but that seems wasteful as J is calculated but not used. @tyronerees any thoughts?

**Comment by AndrewLister-STFC** [comment]
> 
> 
> @AndrewLister-STFC I've spoken to Tyrone and RALFit requires `sum_{i=1}^m (r)_i \nabla^2r_i(x)` passed in as the Hessian. This used to be the case but it got forgotten about when we refactored. Now we could follow the cost functions on this and pass in:
> 
> ```
> r = self.cost_func.eval_r(params, **kwargs)
> H, J = self.cost_func.hes_res(params, **kwargs)
> return np.matmul(H, r))
> ```
> 
> but that seems wasteful as J is calculated but not used. @tyronerees any thoughts?

So RALFit adds the `matmul(2*J.T,J)` component and doesn't work with Poisson, is that right?

**Comment by jfowkes** [comment]
> > @AndrewLister-STFC I've spoken to Tyrone and RALFit requires `sum_{i=1}^m (r)_i \nabla^2r_i(x)` passed in as the Hessian. This used to be the case but it got forgotten about when we refactored. Now we could follow the cost functions on this and pass in:
> > ```
> > r = self.cost_func.eval_r(params, **kwargs)
> > H, J = self.cost_func.hes_res(params, **kwargs)
> > return np.matmul(H, r))
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > but that seems wasteful as J is calculated but not used. @tyronerees any thoughts?
> 
> So RALFit adds the `matmul(2*J.T,J)` component and doesn't work with Poisson, is that right?

Yes I believe so, @tyronerees can confirm regarding Poisson.

**Comment by jfowkes** [comment]
@AndrewLister-STFC ready for another (hopefully final) review round! Key thing to check is that RALFit is happy. Also note that I had to disable the `too-many-lines` pylint warning for `test_controllers.py` as I'm not planning on refactoring the controller tests in this PR.

**Comment by AndrewLister-STFC** [comment]
All looks good to me. Guess I should run it in the morning before merging

**Review by AndrewLister-STFC — APPROVED** [review]

**Pull request merged**